File: scala/src/test/scala/com/laamella/code_state_machine/AutomaticFiringTests.scala
package com.laamella.code_state_machine

import com.laamella.code_state_machine.StateMachineAssert._
import com.laamella.code_state_machine.builder.DslStateMachineBuilder
import com.laamella.code_state_machine.io.DotOutput
import com.laamella.code_state_machine.priority.LeveledPriority
import com.laamella.code_state_machine.priority.LeveledPriority.NORMAL
import com.laamella.code_state_machine.util._
import grizzled.slf4j.Logging
import org.scalatest.Assertions._

class AutomaticFiringTests extends UnitSpec with Logging {
  val builder = new DslStateMachineBuilder[SimpleState, Object, LeveledPriority.Value](NORMAL)
  builder.state(A).isAStartState.always.goTo(B)
  builder.state(B).always.goTo(C)
  builder.state(C).always.goTo(D)
  builder.state(D).always.goTo(E)
  builder.state(E).isAnEndState

  val machine = builder.build()

  trace("\n" + new DotOutput[SimpleState, Object, LeveledPriority.Value]().getOutput(machine))

  "an automatically firing transition" should "fire automatically" in {
    assertActive(machine, A)
    machine.poll()
    assertActive(machine)
  }

}

File: java/src/test/java/com/laamella/code_state_machine/AutomaticFiringTests.java
package com.laamella.code_state_machine;

import com.laamella.code_state_machine.builder.DslStateMachineBuilder;
import com.laamella.code_state_machine.io.DotOutput;
import com.laamella.code_state_machine.priority.Priority;
import com.laamella.code_state_machine.util.SimpleState;
import org.junit.Before;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static com.laamella.code_state_machine.StateMachineAssert.assertActive;
import static com.laamella.code_state_machine.util.SimpleState.*;

public class AutomaticFiringTests {
	private static final Logger log = LoggerFactory.getLogger(AutomaticFiringTests.class);

	private StateMachine<SimpleState, Object, Priority> machine;

	@Before
	public void before() {
		machine = new DslStateMachineBuilder<SimpleState, Object, Priority>(Priority.NORMAL) {
			@Override
			protected void executeBuildInstructions() {
				state(A).isAStartState().when(always()).then(B);
				state(B).when(always()).then(C);
				state(C).when(always()).then(D);
				state(D).when(always()).then(E);
				state(E).isAnEndState();
			}
		}.build(new StateMachine<>());
		log.trace("\n" + new DotOutput<SimpleState, Object, Priority>().getOutput(machine));
	}

	@Test
	public void automaticFirings() {
		assertActive(machine, A);
		machine.poll();
		assertActive(machine);
	}

}
